File: app/src/main/java/com/theappsolutions/boilerplate/data/data_sources/local/DataRepositoryImpl.java
package com.theappsolutions.boilerplate.data.data_sources.local;

import com.theappsolutions.boilerplate.data.model.realm.CachedProject;
import com.theappsolutions.boilerplate.util.storage_utils.RealmUtils;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.realm.OrderedRealmCollection;

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 */
public class DataRepositoryImpl implements DataRepository {

    private RealmManager realmManager;

    public DataRepositoryImpl(RealmManager realmManager) {
        this.realmManager = realmManager;
    }


    @Override
    public Observable<Boolean> addProjects(List<CachedProject> projects) {
        return realmManager.doTransaction(realm -> {
            realm.copyToRealmOrUpdate(projects);
            return true;
        });
    }

    @Override
    public Observable<OrderedRealmCollection<CachedProject>> getProjects() {
        return realmManager.doWithRealmInSpecificThread(realm -> realm.where(CachedProject.class).findAllSorted("id"),
                AndroidSchedulers.mainThread());
    }

    @Override
    public void clearData() {
        RealmUtils.clearRealmTable(CachedProject.class);
    }


}

File: app/src/main/java/com/theappsolutions/boilerplate/data/datasources/local/DataRepositoryImpl.kt
package com.theappsolutions.boilerplate.data.datasources.local

import com.theappsolutions.boilerplate.data.model.realm.CachedProject
import com.theappsolutions.boilerplate.util.storage.RealmUtils

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function
import io.realm.OrderedRealmCollection
import io.realm.Realm

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 */
class DataRepositoryImpl(private val realmManager: RealmManager) : DataRepository {

    override fun getProjects(): Observable<OrderedRealmCollection<CachedProject>> {
        return realmManager.doWithRealmInSpecificThread(Function { realm ->
            realm.where(CachedProject::class.java)
                .findAllSorted("id") },
                AndroidSchedulers.mainThread())
    }

    override fun clearData() {
        RealmUtils.clearRealmTable(CachedProject::class.java)
    }


    override fun addProjects(projects: List<CachedProject>): Observable<Boolean> {
        return realmManager.doTransaction(Function { realm ->
            realm.copyToRealmOrUpdate(projects)
            true
        })
    }
}
